test/test_pebble.py

```diff
@@ -0,0 +1,512 @@
+        self.assertEqual(change.tasks[0].progress.done, 1)
+        self.assertEqual(change.tasks[0].progress.label, '')
+        self.assertEqual(change.tasks[0].progress.total, 1)
+        self.assertEqual(change.tasks[0].ready_time,
+                         datetime_nzdt(2021, 1, 28, 14, 37, 3, 270218))
+        self.assertEqual(change.tasks[0].spawn_time,
+                         datetime_nzdt(2021, 1, 28, 14, 37, 2, 247158))
+        self.assertEqual(change.ready, True)
+        self.assertEqual(change.err, None)
+        self.assertEqual(change.ready_time, datetime_nzdt(2021, 1, 28, 14, 37, 4, 291517))
+        self.assertEqual(change.spawn_time, datetime_nzdt(2021, 1, 28, 14, 37, 2, 247202))
+
+    def test_get_changes(self):
+        empty = {
+            "result": [],
+            "status": "OK",
+            "status-code": 200,
+            "type": "sync"
+        }
+        self.api.responses.append(empty)
+        changes = self.api.get_changes()
+        self.assertEqual(changes, [])
+
+        self.api.responses.append(empty)
+        changes = self.api.get_changes(select=pebble.ChangeState.ALL)
+        self.assertEqual(changes, [])
+
+        self.api.responses.append(empty)
+        changes = self.api.get_changes(select=pebble.ChangeState.ALL, service='foo')
+        self.assertEqual(changes, [])
+
+        self.api.responses.append({
+            "result": [
+                self.get_mock_change_dict(),
+            ],
+            "status": "OK",
+            "status-code": 200,
+            "type": "sync"
+        })
+        changes = self.api.get_changes()
+        self.assertEqual(len(changes), 1)
+        self.assert_mock_change(changes[0])
+
+        self.assertEqual(self.api.requests, [
+            ('GET', '/v1/changes', {'select': 'in-progress'}, None),
+            ('GET', '/v1/changes', {'select': 'all'}, None),
+            ('GET', '/v1/changes', {'select': 'all', 'for': 'foo'}, None),
+            ('GET', '/v1/changes', {'select': 'in-progress'}, None),
+        ])
+
+    def test_get_change(self):
+        self.api.responses.append({
+            "result": self.get_mock_change_dict(),
+            "status": "OK",
+            "status-code": 200,
+            "type": "sync"
+        })
+        change = self.api.get_change('70')
+        self.assert_mock_change(change)
+        self.assertEqual(self.api.requests, [
+            ('GET', '/v1/changes/70', None, None),
+        ])
+
+    def test_abort_change(self):
+        self.api.responses.append({
+            "result": self.get_mock_change_dict(),
+            "status": "OK",
+            "status-code": 200,
+            "type": "sync"
+        })
+        change = self.api.abort_change('70')
+        self.assert_mock_change(change)
+        self.assertEqual(self.api.requests, [
+            ('POST', '/v1/changes/70', None, {'action': 'abort'}),
+        ])
+
+    def _test_services_action(self, action, api_func, services):
```

Comment:
Am I correct in understanding that _test_services_action is testing if we ask for a synchronous change to the API that we are testing that we respond appropriately to the request being accepted but not ready yet? And if we are testing async, then we just set timeout to 0 and assume that we don't check on the change being ready?

These functions feel close to building a fake apiserver for us to engage with, and I'm wondering if that would be clearer if it was cohesively its own testing object vs being snippets spread across a couple of different tests.